pandas/tseries/tdi.py

```diff
@@ -344,6 +344,17 @@ def _add_datelike(self, other):
             other = Timestamp(other)
             i8 = self.asi8
             result = i8 + other.value
+            # gh-14068: there is the possibility of addition overflow,
+            # which occurs when we add two very large positive numbers,
+            # resulting in a negative number.
+            #
+            # Since Timestamp objects can never have a negative value,
+            # we just need to check that the elements in result are also
+            # nonnegative when their corresponding elements in i8 are.
+            for x, y in zip(i8, result):
```

Comment:
I don't think this is performant - shouldn't this really be done via a `checked_add` loop in cython (something like the approach [here](http://blog.regehr.org/archives/1139))?  Then you could use the same elsewhere. 

Might also need to be an option to turn off the checking?  (e.g. do we also check on plain `int64`s)?